File: wrappers/python/aegis_interfaces/messaging/destination.py
"""
Destination abstraction for Pub/Sub messaging.

A Destination represents a complete messaging endpoint including:
- Topic name (where messages are published)
- Subscription mappings (where messages are consumed from)
- Semantic metadata (name, schema, event type)

All destinations are immutable and thread-safe.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Dict, Optional


class EventType(Enum):
    """Type of event carried by a messaging destination."""
    
    # Events from domain topics
    SPECIFICATIONCREATEDEVENTV1 = "aegis-test.specification.created"
    SPECIFICATIONREQUESTEDEVENTV1 = "aegis-test.specification.requested"


@dataclass(frozen=True)
class Destination:
    """
    Immutable representation of an event-driven Pub/Sub messaging destination.
    
    Attributes:
        name: Semantic name in kebab-case (e.g., "specification-created")
        topic: Full Pub/Sub topic name (e.g., "aegis-test.specification.created")
        subscriptions: Mapping of consumer name to subscription name
        event_type: Type of event
        schema: Schema identifier (e.g., "SpecificationCreatedEventV1")
        default_consumer: Optional default consumer for single-consumer topics
    """
    
    name: str
    topic: str
    subscriptions: Dict[str, str]
    event_type: EventType
    schema: str
    default_consumer: Optional[str] = None
    
    def get_subscription(self, consumer: str) -> str:
        """
        Get the subscription name for a specific consumer.
        
        Args:
            consumer: Name of the consuming service
            
        Returns:
            Full subscription name
            
        Raises:
            KeyError: If the consumer is not registered for this topic
        """
        if consumer not in self.subscriptions:
            raise KeyError(
                f"Unknown consumer '{consumer}' for topic '{self.name}'. "
                f"Valid consumers: {list(self.subscriptions.keys())}"
            )
        return self.subscriptions[consumer]
    
    @property
    def subscription(self) -> str:
        """
        Get the default subscription name.
        
        Returns:
            Default subscription name
            
        Raises:
            ValueError: If there's no default consumer
        """
        if self.default_consumer is None:
            raise ValueError(
                f"Topic '{self.name}' has multiple consumers and no default. "
                f"Available consumers: {list(self.subscriptions.keys())}. "
                f"Use get_subscription(consumer) instead."
            )
        return self.get_subscription(self.default_consumer)
    
    def __str__(self) -> str:
        return (
            f"Destination(name='{self.name}', "
            f"topic='{self.topic}', "
            f"schema='{self.schema}')"
        )

File: wrappers/java/src/main/java/com/interfaces/aegis/test/messaging/Destination.java
package com.interfaces.aegis.test.messaging;

/**
 * Represents a complete Pub/Sub messaging destination.
 * 
 * A destination encapsulates:
 * - The topic name (where messages are published)
 * - The subscription name (where messages are consumed from)
 * - Semantic information about purpose and participants
 * 
 * This is an immutable value object that serves as the single source
 * of truth for Pub/Sub routing within Aegis Test.
 * 
 * @implNote All implementations MUST be immutable and thread-safe.
 */
public interface Destination {
    
    /**
     * Returns the semantic name of this destination.
     * Used for logging and debugging purposes.
     * 
     * @return semantic name in kebab-case (e.g., "specification-create-requested")
     */
    String getName();
    
    /**
     * Returns the full Pub/Sub topic name.
     * This is the actual GCP Pub/Sub topic identifier.
     * 
     * @return topic name (e.g., "aegis-test.specification.create.requested")
     */
    String getTopic();
    
    /**
     * Returns the full Pub/Sub subscription name for a specific consumer.
     * 
     * @param consumer the service consuming from this topic
     * @return subscription name (e.g., "orchestrator.aegis-test.specification.create.requested")
     */
    String getSubscription(String consumer);
    
    /**
     * Returns the default subscription name.
     * This is typically used when there's only one primary consumer.
     * 
     * @return default subscription name
     * @throws UnsupportedOperationException if there's no default consumer
     */
    String getSubscription();
    
    /**
     * Returns the schema identifier for the payload.
     * 
     * @return schema name (e.g., "SpecificationCreatedEventV1")
     */
    String getSchema();
}
